### ISSUE TITLE
Potential Problem.

### ISSUE THREAD

**Issue by numairmansur** [numairmansur/Error-Localization#10]
Consider the following program:
```
procedure foo()
{
  var y: int;
  var x: int; 
  var z: int; 
  x := 1;
  y := 2;
  z := 3;
  havoc z;
  assert(z > 10);
}

```
With the following error trace.
` "x := 1;",  "y := 2;", "z := 3;" , "havoc z;" , "assume !(z > 10);" `

Below i show the error trace with it's weakest precondition list.
```
false             false               false               false             (z > 10)                       false
        "x := 1;"         "y := 2;"          "z := 3;"           "havoc z;"             "assume !(z > 10);" 

```

**Output on Ultimate**
```
[L9]         -  x := 1;
       VAL      [x=1]
[L10]        -  y := 2;
       VAL      [x=1, y=2]
[L11]        -  z := 3;
       VAL      [x=1, y=2, z=3]
[L12]        @  havoc z;
       VAL      [x=1, y=2]
[L13]        -  assert z > 10;
       VAL      [x=1, y=2, z=0]
```

**Comment by numairmansur** [comment]
As expected from the behaviour of [WP() for Havoc](https://github.com/numairmansur/Error-Localization/wiki/Weakest-Precondition-operator-WP()) `WP(z>10; havoc z)` returns `false`. 
and hence all the assignment statements in the trace, before havoc are not relevant.

**Comment by numairmansur** [comment]
The implementation works as it should. I think we need to modify the definition of relevancy a little in our original document. 
Form the document, we define restrictiveness and relevancy as follows:
![image](https://cloud.githubusercontent.com/assets/4897599/26146951/3a51e906-3af2-11e7-9ae8-a09754ac368f.png)

But i think we should mention that the assignment statement is relevant if replacing this assignment with havoc causes some assume statement to become restrictive which was not restrictive in the original trace.

**Comment by numairmansur** [comment]
The new definition of Relevancy should be:
![image](https://cloud.githubusercontent.com/assets/4897599/26150828/90439ab8-3b00-11e7-821c-7a110ba54a0c.png)

**Comment by anatskiy** [comment]
test comment

**Comment by Heizmann** [comment]
Thanks for the detailed analysis. This is definitely a step in the right direction, but it is not sufficient.
I added an [example](https://github.com/ultimate-pa/ultimate/commit/aac108ffc4565e646d5c6a990f3e4d77e51ad635#diff-49e2a837f0882e70310637d9a835fb07) to demonstrate the problem.
We probably need something to measure the "amount" of restrictiveness. Maybe we can do this by defining something like a "blocking execution" and by comparing the set of blocked executions before and after replacing assignments by havoc.

**Comment by numairmansur** [comment]
Error Trace
` "y := 42;" "havoc x;" "assume x >= 0 && y >= 23;" "assume !false;" `

Output on Ultimate:
```
[L16]        *  y := 42;
       VAL      [y=42]
[L17]        @  havoc x;
       VAL      [y=42]
[L18]        -  assume x >= 0 && y >= 23;
       VAL      [x=0, y=42]
[L19]        -  assert false;
       VAL      [x=0, y=42]
```

**Comment by numairmansur** [comment]
Hi @Heizmann and @schillic 
I was working on a formal definition of a **blocking execution** and a **set of blocking executions**. But a problem hit my mind. It is possible that the error trace might already have infinite blocking executions before we replace the assignment with havoc. After replacing the assignment with havoc we might have more blocking executions but it will still be infinite. in that case does it make sense to compare the two sets of infinite blocking executions ?
Because i was thinking to declare **x:=t** relevant if after replacing **x:=t** with **havoc**, the set of blocking executions before the replacement is smaller then the set after.

**Comment by schillic** [comment]
Why would infinity be a problem?
Just do not say "|A| < |B|" but say "A is a strict subset of B" ("A \subsetneq B").

**Comment by numairmansur** [comment]
Ok yes that makes sense. I don't know why i was thinking in terms of the size of the two sets.

**Comment by numairmansur** [comment]
Hi @Heizmann @schillic  
What do you think of the new definition for relevancy?
Please let me know if you think something is missing or not making sense. Or if something requires more clarification.
![image](https://cloud.githubusercontent.com/assets/4897599/26522010/3303944e-42f7-11e7-880c-88086b6e7a75.png)

**Comment by Heizmann** [comment]
This is conceptually very good, I think this is exactly the definition that we need. (I am sometimes too optimistic, Christian is much better in finding Problems.)

Maybe the term _blocking executions_ is misleading. Should we call it _blocked execution_?

Let \pi' represent the trace (not a trace).

Since this is an important definition, we should be more verbose here. E.g., as follows.
We say that the assignment statement \pi[i] is relevant if the trace after the replacement has strictly more blocked executions than the trace before the replacement, i.e., if \beta \subsetneq \beta'
(Use \subsetneq instead of \subset)

**Comment by schillic** [comment]
I agree with @Heizmann with a few extra comments (not all of them important).

1. In the intro the index of the statement fits the index of the predecessor state (i.e., you have triples _(s\_i, tf(pi[i]), s\_{i+1})_).
However, in the definition **Blocking Execution** you say _s\_{j-1} does not imply guard(pi[j])_.
Is this intentional? Then I do not understand it.

2. If the above was a typo and _s\_{j-1}_ should be _s\_j_, then there is a contradiction between _s\_j, s\_{j+1} |= tf(pi[j])_ (intro) and _s\_j does not imply guard(pi[j])_ (Blocking Execution) because _tf(pi[j])_ implies _guard(pi[j])_.

3. I think we can keep the definition of a **Blocking Execution** on a more general level and not talk about assume statements at all but instead only talk about transition formulas.
In a simple language the only transition formula that has a nontrivial condition is the assume statement. But in the general case in Ultimate we have statements that are a combination of assume, assignment, and havoc statements and the definition could still make sense.

4. I would not use lowercase letters for sets (epsilon, beta), even with Greek letters. (The capital version of epsilon and beta look like the Latin letter.)
Moreover, I would prefer to turn the sets into functions that take the trace as an argument. In that case I would rather rename them to exec(pi) and block(pi) (or Exec/Block).

**Comment by Heizmann** [comment]
Regarding 3.
I think that lifting all these definitions from statements to transition relations (the TransFormulas in Ultimate) is a separate task. That task should be done for all our definitions and theorems, but I think we also want to have a version of all defintions/theorems for statements.
(We do not want both versions in the paper, but I think we need both versions in order to understand better what we are doing)

**Comment by numairmansur** [comment]
Hi @schillic , regarding your concerns in
## 1.
Yes you are right, it should be s_j. In the following example, in the case of j = 4, for \pi[4], s_j to s_ {j+1} should be  s_4 to s_5. 

![image](https://cloud.githubusercontent.com/assets/4897599/26531376/24e92e72-43e8-11e7-8532-e8edb8c1480d.png)
## 2
I see what you mean here. I have modified the definition a little, what do you think about it now.

![image](https://cloud.githubusercontent.com/assets/4897599/26540830/876e0606-4453-11e7-8157-8dd09f746d5f.png)
## 3
Noted. I am adding this as a separate new task. 
## 4
But the reason i used sets is because i wanted to say in the end that the assignment is relevant if \beta \subsetneq \beta'.  Why do you think turning the sets into functions would be better ?

**Comment by schillic** [comment]
2. Yes, it looks good now.

4. What I meant was a function that takes a trace and produces a set of executions.
This is only a matter of taste, it is equivalent saying "let epsilon bei the executions of pi" and "epsilon(pi)".

**Comment by schillic** [comment]
### Proposed definition of relevancy in case of havoc
havoc(x) is relevant if there exists an assignment x:=t, such that if we replace that havoc with that assignment then all executions of \pi' are blocking (\pi' is infeasible).

**Comment by numairmansur** [comment]
## Still something wrong with the definition.
Hi @Heizmann @schillic,
So, here is the example on which our new definition of relevancy does not work.
`y :=10 ; havoc(x) ; assume(x>10)`
So, if we consider the statement  `y:=10` ,according to our definition it would be relevant because changing it to havoc gives us more blocking executions. But obviously it should not be relevant here.
Please note that if we check it formally with our mathematical formulas and our implementation then it is not relevant. The problem arise in this example only because of how we are defining relevancy with the help of number of blocking executions.

**Comment by numairmansur** [comment]
## New definition
Hi @Heizmann @schillic
What do you think ?
![image](https://user-images.githubusercontent.com/4897599/27259434-9ce8fe4e-5413-11e7-9270-617ff180d25d.png)

**Comment by Heizmann** [comment]
I think the term *infeasible execution* is not good. Either there is an execution or there is no execution. I think the old definition of a *blocked execution* is ok.
I need more time to think about Definition 3.

**Comment by numairmansur** [comment]
Ok. I changed the term "infeasible" with "blocked" in the above comment :)

**Comment by schillic** [comment]
I think the point of @Heizmann was to use the old definition.

About Def. 3:
1. T is undefined here.
2. I would not write _s'\_i+1 \in Q'_ because this is too close to our proof and we do not need _Q'_ here.
3. The definition does not work in the presence of other havocs after the assignment statement. The reason is that some executions might have been blocking even before inserting the havoc. We need to say something like "every execution starting from _s'\_i+1_ is blocking", which (I think) is equivalent to _s'\_i+1 \in Q' \ Q_.

**Comment by numairmansur** [comment]
Hi @schillic 
**Regarding point 3.** Yes that is no problem. Those blocking executions are still there and it is no problem if there already are blocking executions. What we are interested in is the new blocking execution(s) which was introduced because there is a new state in Q' \ Q which is now reachable from P if we change x:=t to havoc(x).

**Comment by numairmansur** [comment]
![states](https://user-images.githubusercontent.com/4897599/27171038-6c46d05a-51b0-11e7-9119-b38537a210eb.jpg)

**Comment by schillic** [comment]
You never said that the new state is in _Q' \ Q_.
But more importantly, the definition is too complicated if we write it like this.

**Comment by numairmansur** [comment]
Oh yes you are right. Ok i am updating it.
![image](https://user-images.githubusercontent.com/4897599/27259869-4068779e-541d-11e7-9722-430f7296297a.png)

**Comment by numairmansur** [comment]
But are you really sure that every execution that start in state  s'_i+1 which is in (Q' \ Q) will be blocking ? 
I think not every execution will be blocking but there exists at least one state for which we will get a blocking execution.

**Comment by numairmansur** [comment]
It is possible that every execution might be blocking but i think we cannot say that this will **always** be the case. I think even if there is one more blocking execution, then that is good enough for us to declare the assignment statement relevant.

**Comment by schillic** [comment]
Q is the error precondition of the suffix trace. That means all states not in Q have only blocking executions for that trace.

**Comment by numairmansur** [comment]
Oh yes i see. You are right.

**Comment by schillic** [comment]
Another version of the definition that @Heizmann and I were discussing just now:

Let pi = "st_1, ..., st_i-1, x := t, st_i+1, ..., st_n" be an error trace of length n.

The assignment at position i is relevant if there is an execution s_1, ..., s_n+1 of pi and some value v such that every execution of the trace "x := v, st_i+1, ..., st_n" starting in s_i is blocking.

**Comment by schillic** [comment]
And a generalization that also works for the havoc case:

Let pi = "st_1, ..., st_n" be an error trace of length n where st_i is an assignment statement or a havoc statement that assigns a new value to variable x.

The statement at position i is relevant if there is an execution s_1, ..., s_n+1 of pi and some value v such that every execution of the trace "x := v, st_i+1, ..., st_n" starting in s_i is blocking.

**Comment by numairmansur** [comment]
Hi @schillic, in the definition

> The assignment at position i is relevant if there is an execution s_1, ..., s_n+1 of pi and some value v such that every execution of the trace "x := v, st_i+1, ..., st_n" starting in s_i is blocking.

Why do we have to say  **_if there is an execution s_1, ..., s_n+1 of pi_**  . Do you mean here that the error trace should be feasible ?

**Comment by schillic** [comment]
Even more: "havoc(x); assume(x == 1)" is feasible, but we are interested in an execution that "goes through".

**Comment by numairmansur** [comment]
I see. 
is it also the same if i replace the last line as following:
"The assignment at position i is relevant if there is an execution s_1, ..., s_n+1 of pi and some value v such that _**the trace "x := v, st_i+1, ..., st_n" is infeasible.**_ "
Am i wrong in thinking that for the trace `x := v, st_i+1, ..., st_n`, it dosen't matter from which state we begin the execution, we will always get a blocking execution ?

**Comment by schillic** [comment]
No, the state s_i matters (not for x but for the other variables).
Take the trace "y := 0; x := 1; assume(x != y)".
After replacing x := 1 by x := 0 the whole trace is infeasible.
But the subtrace "x := 0; assume(x != y)" is still feasible, depending on the state where you start.

**Comment by numairmansur** [comment]
oh yes that is a good example ! ok now i think i also agree with the definition. 
So i will write it down and modify the proof and let you know if i have questions :)

**Comment by numairmansur** [comment]
So here is the [updated document](http://numairmansur.github.io/Error-Localization/temporary/june20/temporary.pdf).
Can you take a look at the proof :)

**Comment by anilv2k** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://link.bnbreward.online/0mhPJ2)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://link.bnbreward.online/0mhPJ2) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @alpaca121, @Kindhearted57, @52, @Olatunji90, @bluekornchips, @randombit, @dipaksaraf, @gunsay, @Roast-Lord, @Adrianf23, @maximebrugel, @xbucd, @Christabel58, @gjeanmart, @knucklesuganda, @ukexim13, @Ciara13122, @JoranHonig, @ottodevs, @mikesdsd, @proger, @vuvoth, @kevinnguyenai, @johnrjj, @TobiasBK, @vforvasilis, @quiz42, @yglsaltfish, @Thinker024, @imwaihon, @Amovane, @SunJ3t, @AkiraMAEGAWA, @SaysZEN, @txemaromero, @dzarezenko, @uduse, @notatestuser, @0xBadMad, @shero4, @candigazuberi2, @lanastasov, @sbreatnach, @Akityo, @0x3f97, @ps1dr3x, @blueplanet, @rl7x, @rootux, @haleembanda
